### PR TITLE
Support for SAML 2.0 asynchronous (front-channel) binding for single logout

### PR DESCRIPTION
**Pull request by sachiniWettasinghe**
### **Proposed changes in this pull request**
* Add HTTP Redirect Binding for front-channel single logout.
* Add HTTP POST Binding for front-channel single logout.

Fixes https://github.com/wso2/product-is/issues/4038

### CODE REVIEW

File: components/org.wso2.carbon.identity.sso.saml.common/src/main/java/org/wso2/carbon/identity/sso/saml/common/Util.java [hunk 1]
```diff
@@ -1,238 +0,0 @@
...
-import java.net.URI;
-import java.net.URISyntaxException;
-import java.net.URL;
-import java.net.URLDecoder;
-import java.util.ArrayList;
-import java.util.HashSet;
-import java.util.List;
-import java.util.Set;
-
-public class Util {
```
Comment (madurangasiriwardena) [inline components/org.wso2.carbon.identity.sso.saml.common/src/main/java/org/wso2/carbon/identity/sso/saml/common/Util.java]: IMO without releasing a major version, we cannot remove a public API.

File: components/org.wso2.carbon.identity.sso.saml/src/main/java/org/wso2/carbon/identity/sso/saml/admin/FileBasedConfigManager.java [hunk 2]
```diff
@@ -158,6 +159,20 @@ public void addServiceProviders() {
                         .fillURLPlaceholders(getTextValue(elem, SAMLSSOConstants.FileBasedSPConfig.SLO_REQUEST_URL)));
             }
 
+            if ((getTextValue(elem, SAMLSSOConstants.FileBasedSPConfig.FRONT_CHANNEL_LOGOUT)) != null) {
+                spDO.setDoFrontChannelLogout(Boolean.valueOf(getTextValue(elem, SAMLSSOConstants.FileBasedSPConfig
+                        .FRONT_CHANNEL_LOGOUT)));
+                if(spDO.isDoFrontChannelLogout()) {
```
Comment (madurangasiriwardena) [inline components/org.wso2.carbon.identity.sso.saml/src/main/java/org/wso2/carbon/identity/sso/saml/admin/FileBasedConfigManager.java]: Formatting issue.

File: components/org.wso2.carbon.identity.sso.saml/src/main/java/org/wso2/carbon/identity/sso/saml/servlet/SAMLSSOProviderServlet.java [hunk 3]
```diff
@@ -764,62 +959,80 @@ private void sendResponse(HttpServletRequest req, HttpServletResponse resp, Stri
...
 
-            if (authenticatedIdPs == null || authenticatedIdPs.isEmpty()) {
-                finalPage = pageWithAcsResponseRelay;
-            } else {
-                finalPage = pageWithAcsResponseRelay.replace(
-                        "<!--$additionalParams-->",
-                        "<input type='hidden' name='AuthenticatedIdPs' value='"
-                                + Encode.forHtmlAttribute(authenticatedIdPs) + "'/>");
-            }
+        PrintWriter out = resp.getWriter();
```
Comment (madurangasiriwardena) [inline components/org.wso2.carbon.identity.sso.saml/src/main/java/org/wso2/carbon/identity/sso/saml/servlet/SAMLSSOProviderServlet.java]: Without writing each and every line to response, it's better to build the whole response and write to output in a single go.

File: components/org.wso2.carbon.identity.sso.saml.stub/src/main/resources/IdentitySAMLSSOConfigService.wsdl [hunk 4]
```diff
@@ -847,4 +849,4 @@
             <http:address location="https://localhost:9443/services/IdentitySAMLSSOConfigService.IdentitySAMLSSOConfigServiceHttpsEndpoint/"/>
         </wsdl:port>
     </wsdl:service>
-</wsdl:definitions>
+</wsdl:definitions>
```
Comment (GDRDABARERA) [inline components/org.wso2.carbon.identity.sso.saml.stub/src/main/resources/IdentitySAMLSSOConfigService.wsdl]: new line please

File: components/org.wso2.carbon.identity.sso.saml.ui/src/main/java/org/wso2/carbon/identity/sso/saml/ui/SAMLSSOUIUtil.java [hunk 5]
```diff
@@ -136,6 +137,42 @@ public static boolean isSingleLogoutEnabled(boolean isSpEdit, SAMLSSOServiceProv
         return false;
     }
 
+    /**
+     * Check front-Channel logout enable and if not enable return false.
+     * @param isSpEdit Operation on service provider, create or edit.
```
Comment (GDRDABARERA) [inline components/org.wso2.carbon.identity.sso.saml.ui/src/main/java/org/wso2/carbon/identity/sso/saml/ui/SAMLSSOUIUtil.java]: Put a space after the description line, to make it clean to read

File: components/org.wso2.carbon.identity.sso.saml.common/src/main/java/org/wso2/carbon/identity/sso/saml/common/Util.java (hunk 1 above)
Comment (sachiniWettasinghe) [reply components/org.wso2.carbon.identity.sso.saml.common/src/main/java/org/wso2/carbon/identity/sso/saml/common/Util.java]: This component was not getting packed in the product. Hence this was removed.

File: components/org.wso2.carbon.identity.sso.saml/src/main/java/org/wso2/carbon/identity/sso/saml/servlet/SAMLSSOProviderServlet.java (hunk 3 above)
Comment (sachiniWettasinghe) [reply components/org.wso2.carbon.identity.sso.saml/src/main/java/org/wso2/carbon/identity/sso/saml/servlet/SAMLSSOProviderServlet.java]: Fixed with 3bbc008dd576ca0169fde3b49fe4f505a99200ea

File: components/org.wso2.carbon.identity.sso.saml/src/main/java/org/wso2/carbon/identity/sso/saml/admin/FileBasedConfigManager.java (hunk 2 above)
Comment (sachiniWettasinghe) [reply components/org.wso2.carbon.identity.sso.saml/src/main/java/org/wso2/carbon/identity/sso/saml/admin/FileBasedConfigManager.java]: Fixed with 1149d9757d9ebea69c35422140a80212a346c963

File: components/org.wso2.carbon.identity.sso.saml.stub/src/main/resources/IdentitySAMLSSOConfigService.wsdl (hunk 4 above)
Comment (sachiniWettasinghe) [reply components/org.wso2.carbon.identity.sso.saml.stub/src/main/resources/IdentitySAMLSSOConfigService.wsdl]: Fixed with 1149d9757d9ebea69c35422140a80212a346c963

File: components/org.wso2.carbon.identity.sso.saml.ui/src/main/java/org/wso2/carbon/identity/sso/saml/ui/SAMLSSOUIUtil.java (hunk 5 above)
Comment (sachiniWettasinghe) [reply components/org.wso2.carbon.identity.sso.saml.ui/src/main/java/org/wso2/carbon/identity/sso/saml/ui/SAMLSSOUIUtil.java]: Fixed with 1149d9757d9ebea69c35422140a80212a346c963

**Pull request merged**